myst_nb/cache.py

```diff
@@ -0,0 +1,192 @@
+
+
+def add_notebook_outputs(env, ntbk, file_path=None):
+    """
+    Add outputs to a NotebookNode by pulling from cache.
+
+    Function to get the database instance. Get the cached output of the notebook
+    and merge it with the original notebook. If there is no cached output,
+    checks if there was error during execution, then saves the traceback to a log file.
+    """
+    # If we have a jupyter_cache, see if there's a cache for this notebook
+    file_path = file_path or env.doc2path(env.docname)
+    dest_path = env.app.outdir
+    reports_dir = dest_path + "/reports"
+    path_cache = False
+
+    if env.config["jupyter_cache"]:
+        path_cache = env.path_cache
+
+    if path_cache:
+        cache_base = get_cache(path_cache)
+        db = cache_base.db
+        cache_record = None
+        r_file_path = _relative_file_path(file_path)
+        try:
+            cache_list = NbCacheRecord.records_from_uri(file_path, db)
+            if len(cache_list):
+                latest = None
+                for item in cache_list:
+                    if latest is None or (latest < item.created):
+                        latest = item.created
+                        latest_record = item
+                cache_record = latest_record
+        except KeyError:
+            cache_record = None
+            logger.error(
+                (
+                    f"Couldn't find cache key for notebook file {r_file_path}. "
+                    "Outputs will not be inserted"
+                )
+            )
+        if cache_record:
+            _, ntbk = cache_base.merge_match_into_notebook(ntbk)
+        else:
+            try:
+                stage_record = cache_base.get_staged_record(file_path)
+            except KeyError:
+                stage_record = None
+            if stage_record and stage_record.traceback:
+                # save the traceback to a log file
+                ensuredir(reports_dir)
+                file_name = r_file_path[
+                    r_file_path.rfind("/") + 1 : r_file_path.rfind(".")
+                ]
+                full_path = reports_dir + "/{}.log".format(file_name)
+                with open(full_path, "w") as log_file:
+                    log_file.write(stage_record.traceback)
+                logger.info(
+                    "Execution traceback for {} is saved in {}".format(
+                        file_name, full_path
+                    )
+                )
+    else:
+        # If we explicitly did not wish to cache, then just execute the notebook
+        ntbk = execute(ntbk)
+
+    return ntbk
+
+
+def execute_staged_nb(cache_base, pk_list):
+    """
+    executing the staged notebook
+    """
+    try:
+        executor = load_executor("basic", cache_base, logger=logger)
+    except ImportError as error:
+        logger.error(str(error))
+        return 1
+    result = executor.run_and_cache(filter_pks=pk_list or None)
+    return result
+
+
+def _relative_file_path(file_path):
```

Comment:
couldn't this just be `Path().relative_to(file_path)` or something close to that?

Reply:
right, have used `r_file_path = Path(file_path).relative_to(Path(file_path).cwd())` instead of this function